source/src_lcao/LCAO_gen_fixedH.cpp

```diff
@@ -1135,6 +1143,214 @@ void LCAO_gen_fixedH::build_Nonlocal_beta_new() //update by liuyu 2021-04-07
     ModuleBase::TITLE("LCAO_gen_fixedH","build_Nonlocal_beta_new");
     ModuleBase::timer::tick ("LCAO_gen_fixedH","build_Nonlocal_beta_new");
 
+#ifdef __MKL
+    const int mkl_threads = mkl_get_max_threads();
+    mkl_set_num_threads(1);
+#endif
+
+    const std::vector<std::vector<std::tuple<int, int, ModuleBase::Vector3<int>, ModuleBase::Vector3<double>>>> adjs_all = GlobalC::GridD.get_adjs();
+
+#ifdef _OPENMP
+    #pragma omp parallel
+    {
+        double* Nonlocal_thread;
+        Nonlocal_thread = new double[GlobalC::ParaO.nloc];
+        ModuleBase::GlobalFunc::ZEROS(Nonlocal_thread, GlobalC::ParaO.nloc);
+        #pragma omp for schedule(dynamic)
+#endif
+        for(int iat=0; iat<GlobalC::ucell.nat; iat++)
+        {
+            const int T0 = GlobalC::ucell.iat2it[iat];
+            const int I0 = GlobalC::ucell.iat2ia[iat];
+            Atom* atom0 = &GlobalC::ucell.atoms[T0];
+
+            //=======================================================
+            //Step1:
+            //saves <beta|psi>, where beta runs over L0,M0 on atom I0
+            //and psi runs over atomic basis sets on the current core
+            //=======================================================
+            #ifdef _OPENMP
+                std::vector<std::unordered_map<int,std::vector<double>>> nlm_tot_thread;
+                nlm_tot_thread.resize(adjs_all[iat].size());
+            #else 
```

Comment:
Sorry but I'm not understand, why OpenMP routine is separated from the non-OpenMP part, with each unique variable?

Reply:
Sorry for the late reply. Considering to be compatible with or without OpenMP compilation option, I use different variable, nlm_tot_thread or nlm_tot. The main difference in using these variables is how to save calculated results to `Hloc_fixed`.  When using OpenMP, there is one more summation of the threads.
`1335         #ifdef _OPENMP
1336             #pragma omp critical(cal_nonlocal)
1337             {
1338                 for(int i=0; i<GlobalC::ParaO.nloc; i++)
1339                 {
1340                     GlobalC::LM.Hloc_fixed[i] += Nonlocal_thread[i];
1341                 }
1342             }
`

Reply:
Sorry for the late reply. Considering to be compatible with or without OpenMP compilation option, I use different variable, `nlm_tot_thread` or `nlm_tot` and other variables. The main difference in using these variables is how to save calculated results to `Hloc_fixed`.  When using OpenMP, there is one more summation of the threads.
```
1335         #ifdef _OPENMP
1336             #pragma omp critical(cal_nonlocal)
1337             {
1338                 for(int i=0; i<GlobalC::ParaO.nloc; i++)
1339                 {
1340                     GlobalC::LM.Hloc_fixed[i] += Nonlocal_thread[i];
1341                 }
1342             }
```

Reply:
Hi @abacus-ustc ,
Thank you for your explanation!
It seems you want to add `Nonlocal_thread` to `GlobalC::LM.Hloc_fixed[i]` for each thread. In this case, you can use OpenMP `reduction` clause to avoid manual sum up these values. 
Here is a minimum example for you:
```c++
//omp.cpp
#include <cstdio>
#include <vector>
#include <cassert>
int main()
{
int n=100;
int* a=new int[n]();

#pragma omp parallel for //reduction(+:a[0:n])
for (int i=0;i<n;i++)
{
for (int j=0;j<n;j++)
a[j]+=i;
}

for (int i=0;i<n;i++)
assert(a[i]==n*(n-1)/2);
//printf("%d\n",a[i]);
return 0;
}
```
Compile it with `g++ omp.cpp -fopenmp`, and you can see it fails without the `reduction` clause.